### PR TITLE
Adds support for modifying allow_pickle and mmap_mode in numpy array type

### PR DESCRIPTION
**Pull request by samhita-alla**
Signed-off-by: Samhita Alla <aallasamhita@gmail.com>

# TL;DR
This PR enables users to specify `allow_pickle` and `mmap_mode` when serializing and deserializing numpy arrays. `allow_pickle` is helpful to serialize and deserialize `dtype=object` numpy arrays. `mmap_mode` is helpful for accessing small fragments of large files without reading the entire file into memory. 

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: flytekit/types/numpy/ndarray.py [hunk 1]
```diff
@@ -11,6 +13,17 @@
 from flytekit.models.types import LiteralType
 
 
+def extract_metadata(t: Type[np.ndarray]) -> Tuple[Type[np.ndarray], Dict[str, bool]]:
+    metadata = None
```
Comment (eapolinario) [inline flytekit/types/numpy/ndarray.py:17]: We can simplify uses of `metadata` if we initialize `metadata` as an empty dictionary (i.e. we won't need to always check if it's non-null).

Comment (samhita-alla) [reply flytekit/types/numpy/ndarray.py]: Ah, you're right!

**Review by eapolinario — APPROVED** [review]
Awesome!
